File: frontend/android/app/src/main/java/com/quietcue/app/data/SmartProfileCoordinator.kt
package com.quietcue.app.data

import com.quietcue.app.domain.PlaceTransition
import com.quietcue.app.domain.SmartProfileDecisionType
import com.quietcue.app.domain.SmartProfilePolicy
import com.quietcue.app.domain.SmartProfileSuggestion
import java.util.UUID
import kotlinx.coroutines.flow.first

enum class SmartTransitionResultType {
    IGNORED,
    SUGGESTED,
    ACTIVATED,
}

data class SmartTransitionResult(
    val type: SmartTransitionResultType,
    val title: String,
    val message: String,
)

class SmartProfileCoordinator(
    private val profileRepository: ProfileRepository,
    private val smartRepository: SmartProfileRepository,
) {
    suspend fun activateManually(profileId: String) {
        profileRepository.setActive(profileId)
        smartRepository.recordManualOverride()
    }

    suspend fun handleTransition(
        placeId: String,
        transition: PlaceTransition,
        nowEpochMs: Long = System.currentTimeMillis(),
        simulated: Boolean = false,
    ): SmartTransitionResult {
        val state = smartRepository.current()
        val place = state.places.firstOrNull { it.id == placeId }
            ?: return ignored("Unknown smart place")
        val catalog = profileRepository.catalog.first()
        val activeProfileId = catalog.activeProfileId
        val decision = SmartProfilePolicy.decide(
            place = place,
            transition = transition,
            state = state,
            activeProfileId = activeProfileId,
            availableProfileIds = catalog.profiles.mapTo(mutableSetOf()) { it.id },
            nowEpochMs = nowEpochMs,
        )
        val targetProfileId = decision.targetProfileId
        val targetProfile = catalog.profiles.firstOrNull { it.id == targetProfileId }
        if (decision.type == SmartProfileDecisionType.IGNORE || targetProfile == null) {
            smartRepository.recordIgnoredTransition(place, transition, nowEpochMs)
            return ignored("No profile change needed")
        }

        if (decision.type == SmartProfileDecisionType.ACTIVATE) {
            profileRepository.setActive(targetProfile.id)
            smartRepository.recordAutomaticActivation(
                place = place,
                transition = transition,
                previousProfileId = activeProfileId,
                targetProfileName = targetProfile.name,
                nowEpochMs = nowEpochMs,
            )
            return SmartTransitionResult(
                type = SmartTransitionResultType.ACTIVATED,
                title = "QuietCue changed profiles",
                message = "${place.name}: ${targetProfile.name} is now active",
            )
        }

        val suggestion = SmartProfileSuggestion(
            id = UUID.randomUUID().toString(),
            placeId = place.id,
            placeName = place.name,
            targetProfileId = targetProfile.id,
            targetProfileName = targetProfile.name,
            transition = transition,
            createdAtEpochMs = nowEpochMs,
            previousProfileId = if (transition == PlaceTransition.ENTER) activeProfileId else null,
            simulated = simulated,
        )
        smartRepository.publishSuggestion(suggestion)
        return SmartTransitionResult(
            type = SmartTransitionResultType.SUGGESTED,
            title = if (transition == PlaceTransition.ENTER) {
                "Profile suggestion for ${place.name}"
            } else {
                "Leaving ${place.name}?"
            },
            message = "Switch to ${targetProfile.name}?",
        )
    }

    suspend fun acceptSuggestion(always: Boolean): SmartProfileSuggestion? {
        val suggestion = smartRepository.current().suggestion ?: return null
        profileRepository.setActive(suggestion.targetProfileId)
        return smartRepository.acceptSuggestion(always)
    }

    suspend fun dismissSuggestion() = smartRepository.dismissSuggestion()

    private fun ignored(message: String) = SmartTransitionResult(
        type = SmartTransitionResultType.IGNORED,
        title = "Smart profiles",
        message = message,
    )
}

File: frontend/ios/QuietCue/Location/SmartProfileCoordinator.swift
import Foundation

enum SmartTransitionResultType {
    case ignored, suggested, activated
}

struct SmartTransitionResult {
    var type: SmartTransitionResultType
    var title: String
    var message: String
}

/// Port of the Android `SmartProfileCoordinator`.
@MainActor
final class SmartProfileCoordinator {
    private let profiles: ProfileStore
    private let smart: SmartProfileStore

    init(profiles: ProfileStore, smart: SmartProfileStore) {
        self.profiles = profiles
        self.smart = smart
    }

    func activateManually(_ profileId: String) throws {
        try profiles.setActive(profileId)
        smart.recordManualOverride()
    }

    func handleTransition(
        placeId: String,
        transition: PlaceTransition,
        nowEpochMs now: Int64 = nowEpochMs(),
        simulated: Bool = false
    ) throws -> SmartTransitionResult {
        let state = smart.state
        guard let place = state.places.first(where: { $0.id == placeId }) else { return ignored("Unknown smart place") }
        let catalog = profiles.catalog
        let activeProfileId = catalog.activeProfileId
        let decision = SmartProfilePolicy.decide(
            place: place,
            transition: transition,
            state: state,
            activeProfileId: activeProfileId,
            availableProfileIds: Set(catalog.profiles.map(\.id)),
            nowEpochMs: now
        )
        guard decision.type != .ignore,
              let targetId = decision.targetProfileId,
              let target = catalog.profiles.first(where: { $0.id == targetId }) else {
            smart.recordIgnoredTransition(place, transition: transition, nowEpochMs: now)
            return ignored("No profile change needed")
        }

        if decision.type == .activate {
            try profiles.setActive(target.id)
            smart.recordAutomaticActivation(
                place,
                transition: transition,
                previousProfileId: activeProfileId,
                targetProfileName: target.name,
                nowEpochMs: now
            )
            return SmartTransitionResult(
                type: .activated,
                title: "QuietCue changed profiles",
                message: "\(place.name): \(target.name) is now active"
            )
        }

        smart.publishSuggestion(
            SmartProfileSuggestion(
                id: UUID().uuidString.lowercased(),
                placeId: place.id,
                placeName: place.name,
                targetProfileId: target.id,
                targetProfileName: target.name,
                transition: transition,
                createdAtEpochMs: now,
                previousProfileId: transition == .enter ? activeProfileId : nil,
                simulated: simulated
            )
        )
        return SmartTransitionResult(
            type: .suggested,
            title: transition == .enter ? "Profile suggestion for \(place.name)" : "Leaving \(place.name)?",
            message: "Switch to \(target.name)?"
        )
    }

    func acceptSuggestion(always: Bool) throws -> SmartProfileSuggestion? {
        guard let suggestion = smart.state.suggestion else { return nil }
        try profiles.setActive(suggestion.targetProfileId)
        return smart.acceptSuggestion(always: always)
    }

    private func ignored(_ message: String) -> SmartTransitionResult {
        SmartTransitionResult(type: .ignored, title: "Smart profiles", message: message)
    }
}
